### DISCUSSION TITLE
Is it better to use useContext hook in app/layout.js?

### DISCUSSION THREAD

**Discussion by naseef0** [vercel/next.js#50146] [App Router]
Instead of passing property to all pages and components we can use context, right? if I use context inside app/layout.js as 

layout.js:

```
export default async function RootLayout({ children, params: { lang },someValue }) {
  return (
    <html lang={lang} dir={lang === "ar" ? "rtl" : "ltr"}>
      <body className={inter.className}>
        <AppState message={someValue}>
            <main>{children}</main>
        </AppState>
      </body>
    </html>
  );
}
```

appState.js
```
"use client";
import AppContext from "./appContext";

export default function AppState({ children,  message}) {
  const value = useMemo(() => ({ message}), [message]);
  return (
    <AppContext.Provider value={value}>
      {children}
    </AppContext.Provider>
  );
}
```

appContext.js:
```
"use client";

import { createContext } from "react";

const AppContext = createContext();

export default AppContext;

```

Will This impact performance?
Will This impact serverside rendering?

**Comment by icyJoseph** [comment]
Let's use some literature as starting point:

- https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves
- https://nextjs.org/docs/getting-started/react-essentials#using-context-in-client-components

> Will This impact performance?

Sort of, in the sense that you'll miss out on the RSC optimizations, and have to pay the price of sending the JavaScript that drives these components, upfront. 

Do remember that once we declare a use client boundary, all direct children are client components too. To be able to slot Server Component children, we have to use composition, where we tell React, "here's a slot for anything", and then when we use the component we put a Server Component on that place. Composition is an old pattern, that not many like, or even bothered to explore. 

- https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props

> "use client" does not need to be defined in every file. The Client module boundary only needs to be defined once, at the "entry point", for all modules imported into it to be considered a Client Component, https://nextjs.org/docs/getting-started/react-essentials#the-use-client-directive

- From the old docs: https://legacy.reactjs.org/docs/composition-vs-inheritance.html

> Will This impact serverside rendering?

Sort of? Not sure what you see as potential impact, since the typical render pass on the server lacks interactivity, and context are usually loaded as empty shells... except perhaps for the case of Themes, and Auth,... in those cases you won't be able to read from the context within a server component, but you could derive the data you need from Cookies or Headers, right?
